### DISCUSSION TITLE
Feature Idea: Ability to create tables from external command outputs.

### DISCUSSION THREAD

**Discussion by sbrow** [nushell/nushell#4109] [Ideas]
nu is very cool. However, IMO, an important feature for a smooth user experience is making it so that external commands output tables just like builtins- without needing to explicitly use `for <format>`

Example: 
It'd be awesome if the following commands had the same result.
<img width="1193" alt="image" src="https://user-images.githubusercontent.com/6729162/140655354-7c6ff9df-50af-4b0f-84c7-67f8129a65f8.png">

Obviously we can't build the output formats for every command into nu- but If there were a way for users to configure the default output format for given commands, I think it would drastically increase the usability of nu. Perhaps something along the lines of:

```toml
# config.toml
[default_formats]
jq = 'from json'
```

the configuration above would insert `| from json` into the pipeline whenever the external command `jq` is executed.
  i.e. `echo '{"foo":"bar"}' | jq | select foo` would expand to `echo '{"foo":"bar"}' | jq | $nu.default_formats.jq | select foo`

It seems the only missing feature to implement such a change would be a way to hook into the pipeline when external commands are called. This would depend on the hook feature, described in issue #3872, and the hook would require access to the pipeline.

Anyway, keep up the good work and let me know what you think! 😄

**Comment by webbedspace** [comment]
I figure you could write something like this in your config:
```
alias jq = (jq | from json)
```
…except that you can't pass arguments to jq this way. (I still need to figure out how to write `def`s that can take variadic input and pass them to a command).

**Reply by sbrow** [reply]
That's why I recommended defining it in the config file. Then all it needs to do is take commands that have a string output, look them up in the config file, and format them appropriately before they're displayed.

**Comment by theAkito** [comment]
I don't understand the need for `jq` in any place, especially the Nu Shell.

**Reply by sbrow** [reply]
> I don't understand the need for `jq` in any place, especially the Nu Shell.

`jq` was simply an example. I could just as easily have used `awk`, `csvkit`, a custom written `perl` or `bash` script, `tcpdump`, or anything else.

**Reply by theAkito** [reply]
@sbrow 

What about pre-execution Hooks?

https://www.nushell.sh/book/hooks.html#hooks

You could write your own definition of looking up whether a substring fits an item in a list and if yes, append that `| from json` to it, or something.

**Reply by sbrow** [reply]
When I first wrote the issue, hooks hadn't been completed yet. Now that they are, I believe that might do the trick. Thanks!
